File: crates/sc-core/src/trash.rs
//! Per-share trash: `<share>/.sctrash/{id}-{encoded_orig_path}`
//! `TrashMode::ShareLocal`. `TrashMode::Off` —
//! the default for every share — is handled directly in `ops::delete` as a
//! plain unlink; this module only runs when a share has trash turned on.
//! An admin toggles it per share (`Core::update_share`'s `trash_enabled`,
//! persisted in `shares.db` — see `share.rs`), not through `config.toml`.
//!
//! `.sctrash` is *flat* — one level, not a mirror of the share's directory
//! tree — so a trashed file's original parent directory has nowhere to live
//! except folded into its own entry name. Earlier this crate kept only
//! `r.path.name()` (the basename) there, which silently discarded the
//! parent: restoring a file trashed from `docs/2024/report.pdf` dropped it
//! at the share root as `report.pdf` with no error and no indication
//! anything had moved. Confirmed by checking the actual on-disk state after
//! a delete+restore round trip, not by reading the code. `encode_orig_path`/
//! `decode_orig_path` below fix that by carrying the *full* relative path
//! through the flat entry name instead of just the leaf.

use sc_acl::Perms;
use sc_vfs::{SafePath, ShareId, ShareRoot, Stat, UserId};

use crate::entry::TrashEntry;
use crate::error::CoreError;
use crate::ops::path_exists;
use crate::resolve::Resolved;

pub(crate) const TRASH_DIR: &str = ".sctrash";

/// Fold a share-relative path into a single filesystem-safe path component.
/// Base64url (no padding) is used rather than the path's own `/`-joined
/// form because `/` inside a *single* path component would be reinterpreted
/// as a separator by `SafePath::join`/the OS — exactly the character this
/// needs to survive intact. The alphabet (`[A-Za-z0-9_-]`) also can't
/// collide with any of `validate_component`'s other rejections (no NUL/
/// control bytes, no `:`, never ends in `.`/` `), so the encoded name always
/// passes the same component validation every other on-disk name does.
fn encode_orig_path(path: &SafePath) -> String {
    data_encoding::BASE64URL_NOPAD.encode(path.to_display_string().as_bytes())
}

/// Inverse of [`encode_orig_path`]. Returns `None` on anything that doesn't
/// decode to a valid `SafePath` — in particular, a pre-fix trash entry whose
/// suffix is a literal basename (not base64) rather than an encoded path.
/// Callers treat that as the legacy shape, not an error: an entry trashed
/// before this fix shipped has no recorded parent directory to recover
/// (it was never written down), so the only honest fallback is the old
/// behavior, restore-to-root, not a hard failure on an otherwise-valid trash
/// entry.
fn decode_orig_path(encoded: &str, max_depth: u16) -> Option<SafePath> {
    let bytes = data_encoding::BASE64URL_NOPAD.decode(encoded.as_bytes()).ok()?;
    let s = String::from_utf8(bytes).ok()?;
    SafePath::parse(&s, max_depth).ok()
}

impl crate::Core {
    pub(crate) fn trash_move(&self, r: &Resolved, _st: &Stat) -> Result<(), CoreError> {
        let max_depth = r.root.policy().max_depth;
        let trash_dir = SafePath::control(TRASH_DIR, max_depth)?;
        if !path_exists(&r.root, &trash_dir)? {
            r.root.mkdir(&trash_dir)?;
        }

        let id = uuid::Uuid::new_v4().simple().to_string();
        // Whole path, not just `r.path.name()` -- see the module doc comment
        // for why the basename alone lost the file's location on restore.
        let encoded = encode_orig_path(&r.path);
        let trash_name = format!("{id}-{encoded}");
        let trash_path = trash_dir.join(&trash_name, max_depth)?;
        r.root.rename(&r.path, &trash_path, true)?;
        Ok(())
    }

    fn split_trash_name(name: &str) -> Option<(&str, &str)> {
        // `id` is `Uuid::simple()` -- 32 lowercase hex chars, never contains
        // `-` -- so splitting on the *first* `-` always separates the id
        // from the (base64url-encoded, itself possibly `-`-containing)
        // remainder correctly, regardless of trash-entry format version.
        name.split_once('-')
    }

    /// Best-effort display name for a trash entry: the leaf of the decoded
    /// original path, or (legacy entries / anything undecodable) the raw
    /// suffix as-is, which pre-fix was already exactly the basename.
    fn trash_display_name(rest: &str, max_depth: u16) -> String {
        match decode_orig_path(rest, max_depth) {
            Some(p) => p.name().unwrap_or(rest).to_string(),
            None => rest.to_string(),
        }
    }

    /// Create every ancestor of `dir` that doesn't already exist, shallowest
    /// first -- `ShareRoot::mkdir` is one level at a time, there is no
    /// recursive `mkdir -p` in that API. Used by `trash_restore` to recreate
    /// a directory that no longer exists (deleted itself, independently of
    /// the file now being restored into it) so the file can go back to
    /// exactly the path it was trashed from.
    ///
    /// Recreating is the chosen behavior over the two alternatives the task
    /// considered: dropping the file at the share root (the original bug —
    /// silent, and exactly what this fix removes) or at the nearest
    /// surviving ancestor (also silent, and now the *file's* name is right
    /// but its parent directory is wrong in a way the response never
    /// mentions). An empty recreated directory is the one outcome a user
    /// restoring a file already expects — it is what "put it back" means —
    /// and it costs nothing extra if the directory turns out to already
    /// exist.
    fn ensure_dir_recursive(root: &ShareRoot, dir: &SafePath, max_depth: u16) -> Result<(), CoreError> {
        let mut built = SafePath::root();
        for comp in dir.components() {
            built = built.join(comp.as_str(), max_depth)?;
            if path_exists(root, &built)? {
                continue;
            }
            match root.mkdir(&built) {
                Ok(()) => {}
                // Lost a race with something else creating the same
                // ancestor (another restore into a sibling path, a fresh
                // upload, ...) -- the directory existing is the only thing
                // that mattered.
                Err(sc_vfs::VfsError::AlreadyExists) => {}
                Err(e) => return Err(e.into()),
            }
        }
        Ok(())
    }

    pub fn trash_list(&self, user: UserId, share: ShareId) -> Result<Vec<TrashEntry>, CoreError> {
        let root = self.share(share).ok_or(CoreError::NotFound)?;
        let decision = self.acl.evaluate(user, share, &SafePath::root(), Perms::READ);
        if !decision.is_allowed() {
            return Err(CoreError::Denied { by: decision.by() });
        }

        let max_depth = root.policy().max_depth;
        let trash_dir = SafePath::control(TRASH_DIR, max_depth)?;
        let entries = root.read_dir(&trash_dir).unwrap_or_default();
        let mut out = Vec::new();
        for e in entries {
            let Some((id, rest)) = Self::split_trash_name(&e.name) else { continue };
            let Ok(p) = trash_dir.join(&e.name, max_depth) else { continue };
            if let Ok(st) = root.stat(&p) {
                out.push(TrashEntry {
                    id: id.to_string(),
                    name: Self::trash_display_name(rest, max_depth),
                    size: st.size,
                    deleted_mtime_ns: st.mtime_ns,
                });
            }
        }
        Ok(out)
    }

    pub fn trash_restore(&self, user: UserId, share: ShareId, id: &str) -> Result<(), CoreError> {
        let root = self.share(share).ok_or(CoreError::NotFound)?;
        let decision = self.acl.evaluate(user, share, &SafePath::root(), Perms::CREATE);
        if !decision.is_allowed() {
            return Err(CoreError::Denied { by: decision.by() });
        }

        let max_depth = root.policy().max_depth;
        let trash_dir = SafePath::control(TRASH_DIR, max_depth)?;
        let entries = root.read_dir(&trash_dir)?;
        let prefix = format!("{id}-");
        let entry_name = entries
            .into_iter()
            .map(|e| e.name.to_string())
            .find(|n| n.starts_with(&prefix))
            .ok_or(CoreError::NotFound)?;
        let rest = Self::split_trash_name(&entry_name)
            .map(|(_, n)| n)
            .ok_or_else(|| CoreError::Internal("malformed trash entry".into()))?;

        // The path this entry was trashed from, recovered whole
        // (`decode_orig_path`) -- or, for a pre-fix entry that has no
        // recorded parent, the legacy fallback of treating `rest` as a bare
        // basename and restoring to the share root exactly as this used to
        // unconditionally do. That fallback is the *documented* old
        // behavior applied only where the new information doesn't exist,
        // not a silent downgrade of a fixed entry.
        let dest_path = match decode_orig_path(rest, max_depth) {
            Some(p) => p,
            None => SafePath::parse(rest, max_depth)?,
        };

        let trash_path = trash_dir.join(&entry_name, max_depth)?;

        // The original directory may itself have been deleted (independent
        // of the file being restored into it) since the file was trashed.
        // Recreate the whole ancestor chain rather than dropping the file
        // somewhere else silently -- see `ensure_dir_recursive`'s doc
        // comment for why recreation was chosen over the alternatives.
        Self::ensure_dir_recursive(&root, &dest_path.parent(), max_depth)?;

        if path_exists(&root, &dest_path)? {
            return Err(CoreError::Conflict);
        }
        root.rename(&trash_path, &dest_path, true)?;
        self.mark_dirty(share, &dest_path);
        Ok(())
    }

    pub fn trash_purge(&self, user: UserId, share: ShareId, id: Option<&str>) -> Result<(), CoreError> {
        let root = self.share(share).ok_or(CoreError::NotFound)?;
        let decision = self.acl.evaluate(user, share, &SafePath::root(), Perms::DELETE);
        if !decision.is_allowed() {
            return Err(CoreError::Denied { by: decision.by() });
        }

        let max_depth = root.policy().max_depth;
        let trash_dir = SafePath::control(TRASH_DIR, max_depth)?;
        let entries = root.read_dir(&trash_dir).unwrap_or_default();
        for e in entries {
            let matches = match id {
                Some(id) => e.name.starts_with(&format!("{id}-")),
                None => true,
            };
            if !matches {
                continue;
            }
            let p = trash_dir.join(&e.name, max_depth)?;
            // Quota charge-back: purge is where trashed
            // bytes are actually freed (`trash_move` only relocated them, no
            // charge there). Size must be read before the delete.
            let freed = if e.kind == sc_vfs::Kind::Dir { self.aggregate(share, &p)?.rsize } else { root.stat(&p).map(|st| st.size).unwrap_or(0) };
            if e.kind == sc_vfs::Kind::Dir {
                Self::delete_recursive(&root, &p)?;
            } else {
                root.unlink(&p)?;
            }
            self.charge_quota(user, -(freed as i64));
        }
        Ok(())
    }
}

File: go/internal/core/trash.go
//go:build linux

package core

import (
	"context"
	"crypto/rand"
	"encoding/base64"
	"errors"
	"fmt"
	"strings"

	"github.com/heavycaffeiner/stowcloud/go/internal/acl"
	"github.com/heavycaffeiner/stowcloud/go/internal/vfs"
)

// Per-share trash: <share>/.sctrash/{id}-{encoded origin path}, one flat
// level. The .sctrash directory is a control directory, so it never appears in
// a listing and is never indexed. Trash is off by default; the toggle is
// visible in the share definition, and until it is on a delete is a delete
// and the UI says so before the destructive action.
//
// The entry name carries the whole origin path, not just the basename, because
// a flat store has nowhere else to keep the parent: restoring a file trashed
// from docs/2024/report.pdf has to land there, not at the share root.
const trashDir = ".sctrash"

// TrashEntry is one entry of the trash listing.
type TrashEntry struct {
	ID string
	// Name is the display name: the leaf of the origin path, or the raw
	// suffix for a legacy entry that predates the encoding.
	Name string
	// OrigPath is the share-relative path the entry was deleted from, empty
	// when the entry predates that encoding and only its basename was kept.
	OrigPath string
	IsDir    bool
	Size     uint64
	// DeletedAtNs is the inode change time the move into the trash set. It
	// used to carry the file's mtime, which a move does not touch, so a file
	// last edited a year ago and deleted a minute ago was listed as "Deleted"
	// a year ago.
	DeletedAtNs int64
}

// trashMove moves one entry into the trash. The id is random so two
// concurrent deletes of the same name cannot collide on the entry name.
func (c *Core) trashMove(ctx context.Context, r Resolved, st vfs.Stat) error {
	trash, err := c.ensureTrashDir(ctx, r)
	if err != nil {
		return err
	}
	var b [8]byte
	if _, rerr := rand.Read(b[:]); rerr != nil {
		return fmt.Errorf("naming a trash entry: %w", rerr)
	}
	id := hexLower(b[:])
	encoded := encodeOrigPath(r.path)
	entryName, err := trash.Join(id + "-" + encoded)
	if err != nil {
		return err
	}
	if err := r.root.Rename(r.path, entryName, true); err != nil {
		return mapVFSErr(err)
	}
	c.markDirty(ctx, r.share, r.path)
	c.markDirty(ctx, r.share, trash)
	return nil
}

// ensureTrashDir returns the share's trash directory, creating it first.
func (c *Core) ensureTrashDir(ctx context.Context, r Resolved) (vfs.SafePath, error) {
	trash, err := vfs.RootPath().JoinControl(trashDir)
	if err != nil {
		return vfs.SafePath{}, err
	}
	if err := r.root.Mkdir(trash); err != nil {
		if !errors.Is(err, vfs.ErrExists) {
			return vfs.SafePath{}, mapVFSErr(err)
		}
	}
	return trash, nil
}

// encodeOrigPath folds a share-relative path into a single filesystem-safe
// component. Base64url is used rather than the path's own slash-joined form
// because a slash inside a single component would be reinterpreted as a
// separator, exactly the character that must survive intact.
func encodeOrigPath(p vfs.SafePath) string {
	return base64.RawURLEncoding.EncodeToString([]byte(p.String()))
}

// decodeOrigPath is the inverse, and reports false on anything that does not
// decode to a valid safe path. A pre-encoding trash entry whose suffix is a
// literal basename is exactly that, and the caller treats it as the legacy
// shape: restore to the share root, which is all the information carried.
func decodeOrigPath(encoded string) (vfs.SafePath, bool) {
	b, err := base64.RawURLEncoding.DecodeString(encoded)
	if err != nil {
		return vfs.SafePath{}, false
	}
	p, err := vfs.ParseSafePath(string(b))
	if err != nil {
		return vfs.SafePath{}, false
	}
	return p, true
}

// splitTrashName separates the id from the encoded origin path. The id is
// lowercase hex and never contains a dash, so splitting on the first dash
// always separates the two even though the encoded half may itself contain
// dashes.
func splitTrashName(name string) (id, rest string, ok bool) {
	id, rest, ok = strings.Cut(name, "-")
	return
}

// trashDisplayName is the leaf of the decoded origin path for the listing,
// or the raw suffix for a legacy entry.
func trashDisplayName(rest string) string {
	if p, ok := decodeOrigPath(rest); ok {
		return p.Name()
	}
	return rest
}

// TrashList returns the share's trash. It requires READ on the share root,
// which is the grant that decides who may see the share at all.
func (c *Core) TrashList(ctx context.Context, r Resolved) ([]TrashEntry, error) {
	if err := r.Require(acl.Read); err != nil {
		return nil, err
	}
	trash, err := vfs.RootPath().JoinControl(trashDir)
	if err != nil {
		return nil, err
	}
	entries, err := r.root.ReadDir(trash, vfs.HideReserved)
	if err != nil {
		if errors.Is(err, vfs.ErrNotFound) {
			return nil, nil
		}
		return nil, mapVFSErr(err)
	}
	out := make([]TrashEntry, 0, len(entries))
	for _, e := range entries {
		id, rest, ok := splitTrashName(e.Name)
		if !ok {
			continue
		}
		entryPath, jerr := trash.JoinExisting(e.Name)
		if jerr != nil {
			continue
		}
		st, serr := r.root.Stat(entryPath)
		if serr != nil {
			continue
		}
		te := TrashEntry{
			ID:          id,
			Name:        trashDisplayName(rest),
			IsDir:       st.Kind.IsDir(),
			Size:        st.Size,
			DeletedAtNs: ctimeOrMtime(st),
		}
		if p, ok := decodeOrigPath(rest); ok {
			te.OrigPath = p.String()
		}
		out = append(out, te)
	}
	return out, nil
}

// ctimeOrMtime is the deletion time: the entry moved into the trash is an
// inode change, which is what records when that happened. A filesystem with no
// change time falls back to the file's own mtime rather than showing nothing.
func ctimeOrMtime(st vfs.Stat) int64 {
	if st.CtimeNs != nil {
		return *st.CtimeNs
	}
	return st.MtimeNs
}

// TrashRestore restores one entry to the path it was trashed from, recreating
// the ancestor chain if the original directories were themselves deleted. A
// conflict is produced rather than an overwrite: something is at the origin
// path now, and the caller has to act on that.
//
// The origin path is resolved again at restore time, which is the rule that
// makes restore safe: a path that no longer resolves the same way produces a
// conflict rather than a silent overwrite.
func (c *Core) TrashRestore(ctx context.Context, r Resolved, id string) (vfs.SafePath, error) {
	if err := r.Require(acl.Create); err != nil {
		return vfs.SafePath{}, err
	}
	trash, err := vfs.RootPath().JoinControl(trashDir)
	if err != nil {
		return vfs.SafePath{}, err
	}
	entries, err := r.root.ReadDir(trash, vfs.HideReserved)
	if err != nil {
		return vfs.SafePath{}, mapVFSErr(err)
	}
	var entryName string
	found := false
	for _, e := range entries {
		if strings.HasPrefix(e.Name, id+"-") {
			entryName = e.Name
			found = true
			break
		}
	}
	if !found {
		return vfs.SafePath{}, ErrNotFound
	}
	_, rest, _ := splitTrashName(entryName)

	var dest vfs.SafePath
	if p, ok := decodeOrigPath(rest); ok {
		dest = p
	} else {
		// Legacy entry: the suffix is a bare basename and was always restored
		// to the share root. That is the documented old behavior, applied only
		// where the new information does not exist.
		p, perr := vfs.ParseSafePath(rest)
		if perr != nil {
			return vfs.SafePath{}, ErrNotFound
		}
		dest = p
	}

	if err := ensureDirRecursive(ctx, r, dest.Parent()); err != nil {
		return vfs.SafePath{}, err
	}
	if p, err := pathExists(r.root, dest); err != nil {
		return vfs.SafePath{}, err
	} else if p {
		return vfs.SafePath{}, ErrConflict
	}

	trashPath, jerr := trash.JoinExisting(entryName)
	if jerr != nil {
		return vfs.SafePath{}, jerr
	}
	if err := r.root.Rename(trashPath, dest, true); err != nil {
		return vfs.SafePath{}, mapVFSErr(err)
	}
	c.markDirty(ctx, r.share, dest)
	c.markDirty(ctx, r.share, trash)
	return dest, nil
}

// ensureDirRecursive creates every ancestor of dir that does not exist,
// shallowest first. ShareRoot.Mkdir is one level at a time, so a directory
// that was itself deleted has to be rebuilt for the restore to land where it
// was trashed from.
func ensureDirRecursive(ctx context.Context, r Resolved, dir vfs.SafePath) error {
	cur := vfs.RootPath()
	for _, comp := range dir.Components() {
		var jerr error
		cur, jerr = cur.JoinExisting(comp)
		if jerr != nil {
			return jerr
		}
		if p, err := pathExists(r.root, cur); err != nil {
			return err
		} else if p {
			continue
		}
		if err := r.root.Mkdir(cur); err != nil {
			// Lost a race with something else creating the same ancestor,
			// which is fine: the directory existing is all that mattered.
			if !errors.Is(err, vfs.ErrExists) {
				return mapVFSErr(err)
			}
		}
	}
	return nil
}

// TrashPurge deletes trash entries, one by name or all of them. It is where
// trashed bytes are actually freed, so the quota ledger is credited here and
// not in trashMove, which only relocated them.
func (c *Core) TrashPurge(ctx context.Context, r Resolved, id *string) error {
	if err := r.Require(acl.Delete); err != nil {
		return err
	}
	trash, err := vfs.RootPath().JoinControl(trashDir)
	if err != nil {
		return err
	}
	entries, err := r.root.ReadDir(trash, vfs.HideReserved)
	if err != nil {
		if errors.Is(err, vfs.ErrNotFound) {
			return nil
		}
		return mapVFSErr(err)
	}
	// A failure to remove one entry is reported rather than skipped. Every
	// branch below used to `continue`, so a purge that deleted nothing still
	// answered success: the screen said the item was gone and the next listing
	// showed it again.
	var failed error
	for _, e := range entries {
		if id != nil && !strings.HasPrefix(e.Name, *id+"-") {
			continue
		}
		entryPath, jerr := trash.JoinExisting(e.Name)
		if jerr != nil {
			failed = errors.Join(failed, jerr)
			continue
		}
		st, serr := r.root.Stat(entryPath)
		if serr != nil {
			failed = errors.Join(failed, mapVFSErr(serr))
			continue
		}
		var freed uint64
		if st.Kind.IsDir() {
			// What the quota is credited, and nothing else. The rollup refuses
			// a path under the trash directory because that name is a control
			// prefix, which made every directory purge stop here: the entry
			// stayed on disk and the caller was told it was deleted. Losing the
			// credit is worth strictly less than losing the delete, so a
			// refusal here costs the ledger and not the operation.
			if agg, aerr := c.Aggregate(ctx, r.share, entryPath); aerr == nil {
				freed = agg.RSize
			}
			sub := Resolved{user: r.user, share: r.share, root: r.root, path: entryPath, perms: r.perms}
			if err := c.deleteRecursive(ctx, sub); err != nil {
				failed = errors.Join(failed, err)
				continue
			}
		} else {
			freed = st.Size
			if err := r.root.Unlink(entryPath); err != nil {
				failed = errors.Join(failed, mapVFSErr(err))
				continue
			}
		}
		if freed > 0 {
			c.chargeQuota(ctx, r.user, int64Minus(freed))
		}
	}
	c.markDirty(ctx, r.share, trash)
	return failed
}

func hexLower(b []byte) string {
	const hex = "0123456789abcdef"
	out := make([]byte, 0, 2*len(b))
	for _, c := range b {
		out = append(out, hex[c>>4], hex[c&0xf])
	}
	return string(out)
}

// int64Minus negates a size for the ledger, saturating at the most negative
// value that fits: a freed size cannot exceed the signed range, and a size
// that would is clamped rather than wrapped.
func int64Minus(v uint64) int64 {
	const max = ^uint64(0) >> 1
	if v > max {
		return -int64(max)
	}
	return -int64(v) //nolint:gosec // v was clamped to the signed range above.
}
